crates/sui-axelar-cgp/move/sources/gateway.move

```diff
@@ -0,0 +1,419 @@
+    /// Does proof validation, fails when proof is invalid or if weight
+    /// threshold is not reached.
+    public fun validate_proof(
+        validators: &Axelar,
+        message_hash: vector<u8>,
+        proof: vector<u8>
+    ): bool {
+        // Turn everything into bcs bytes and split data.
+        let proof = bcs::new(proof);
+        let (operators, weights, threshold, signatures) = (
+            bcs::peel_vec_vec_u8(&mut proof),
+            bcs::peel_vec_u128(&mut proof),
+            bcs::peel_u128(&mut proof),
+            bcs::peel_vec_vec_u8(&mut proof)
+        );
+
+        let operators_length = vector::length(&operators);
+        let operators_epoch = *vec_map::get(
+            &validators.epoch_for_hash,
+            &operators_hash(&operators, &weights, threshold)
+        );
+        let epoch = validators.epoch;
+
+        assert!(operators_epoch != 0 && epoch - operators_epoch < OLD_KEY_RETENTION, EInvalidOperators);
+
+        let (i, weight, operator_index) = (0, 0, 0);
+        let total_signatures = vector::length(&signatures);
+
+        while (i < total_signatures) {
+            let signature = *vector::borrow(&signatures, i);
+            normalize_signature(&mut signature);
+
+            let signed_by: vector<u8> = ecdsa::secp256k1_ecrecover(&signature, &message_hash, 0);
+            while (operator_index < operators_length && &signed_by != vector::borrow(&operators, operator_index)) {
+                operator_index = operator_index + 1;
+            };
+
+            // assert!(operator_index == operators_length, 0); // EMalformedSigners
+
+            weight = weight + *vector::borrow(&weights, operator_index);
+            if (weight >= threshold) { return true };
+            operator_index = operator_index + 1;
+        };
+
+        abort ELowSignaturesWeight
+    }
+
+    public(friend) fun transfer_operatorship(axelar: &mut Axelar, payload: vector<u8>) {
+        let bcs = bcs::new(payload);
+        let new_operators = bcs::peel_vec_vec_u8(&mut bcs);
+        let new_weights = bcs::peel_vec_u128(&mut bcs);
+        let new_threshold = bcs::peel_u128(&mut bcs);
+
+        let operators_length = vector::length(&new_operators);
+        let weight_length = vector::length(&new_weights);
+
+        assert!(operators_length != 0, EInvalidOperators);
+        // TODO: implement `_isSortedAscAndContainsNoDuplicate` function.
+
+        assert!(weight_length == operators_length, EInvalidWeights);
+        let (total_weight, i) = (0, 0);
+        while (i < weight_length) {
+            total_weight = total_weight + *vector::borrow(&new_weights, i);
+            i = i + 1;
+        };
+        assert!(!(new_threshold == 0 || total_weight < new_threshold), EInvalidThreshold);
+
+        // TODO: Can we assume that the new operators hash won't collide with the old ones?
+        let new_operators_hash = operators_hash(&new_operators, &new_weights, new_threshold);
```

Comment:
Can we assume new operators + weights + threshold is always different from the previous ones? If not we will need to handle hash collision.

Reply:
I think it makes sense to handle the hash collision for the best security practice